src/main/scala/org/apache/sysml/compiler/macros/RewriteMacros.scala

```diff
@@ -0,0 +1,218 @@
+          case api.ValRef(sym) => sym
+          case api.VarRef(sym) => sym
+          case api.BindingRef(sym) => sym
+          case api.DefCall(Some(target), method, targs, args) => method
+        })
+        .traverseAny(e.tree)
+    }
+
+    // take the closure and remove all "legal" accesses
+    val closure = defcalls diff valdefs diff inputs
+
+    // construct maps for all inputs to the MLContext and binding references
+    val inputMap = inputs.map(x => x.name.decodedName.toString -> x).toMap
+    val bindingRefMap = bindingRefs.map(x => x.name.decodedName.toString -> x).toMap
+
+    /** extract the return type that has to be retained from mlcontext */
+    val (outType: u.Type, outNames: List[u.Tree]) = e.tree match {
+      case u.Block(_, expr) => expr match {
+        case l: u.Literal if !(expr.tpe =:= u.typeOf[Unit]) => (l.tpe, List(l.value))
+        case a: u.Apply if a.symbol.name == u.TermName("apply") => (a.tpe, a.args)
+        case _ if expr.tpe =:= u.typeOf[Unit] =>
+          (u.typeOf[Unit], List())
+        case _ =>
+          (expr.tpe, List(expr))
+      }
+      case _ =>
+        (e.tree.tpe, e.tree)
+    }
+
+    // generate the actual DML code
+    val dmlString = toDML(dmlPipeline(typeCheck = false)()(e.tree))(new Environment(inputMap, bindingRefMap, 0))
+
+    // prepend line numbers
+    val formatted = dmlString.split("\n").zipWithIndex.map(tup => f"${tup._2 + 1}%4d|${tup._1}").mkString("\n")
+
+    // assemble the input and output parameters to MLContext
+    val inParams  = inputs.map(in => (in.name.decodedName.toString, in))
+    val outParams = outNames.map(_.symbol.name.toString)
+
+    // assemble the type of the return expression we want from MLContext
+    val outTypes  = outType.typeArgs match {
+      case Nil => List(outType)
+      case ls => ls
+    }
+
+    // if the return type is Unit, we don't want to call getTuple
+    val result = if (outParams.isEmpty) q"()" else q"res.getTuple[..${outTypes}](..${outParams})"
+
+    // this is a workaround for the fact that MLContext only returns tuples
+    val out = if (outTypes.length == 1 && outParams.nonEmpty) q"out._1" else q"out"
+
+    // Construct algorithm object
+    val alg = q"""{
+      new _root_.org.apache.sysml.api.linalg.SystemMLAlgorithm[${u.weakTypeOf[T]}]  {
+
+      import _root_.org.apache.sysml.api.mlcontext.ScriptFactory._
+      import _root_.scala.reflect._
+
+      val inputs = Seq(..${inParams})
+      val outputs = Seq(..${outParams})
+
+      def run(ml: _root_.org.apache.sysml.api.mlcontext.MLContext, printDML: Boolean = false): ${u.weakTypeOf[T]} = {
+        if (printDML) {
+          println("=" * 80)
+          println((" " * 26) + "RUNNING GENERATED DML SCRIPT")
+          println("=" * 80)
+          println(${formatted})
+          println("=" * 80)
+          println("Input parameters:" +  List(..${inParams}).mkString(", "))
+        }
+
+        val script = dml($dmlString).in(inputs).out(..${outParams})
```

Comment:
Maybe break this line out as a method of the superclass so that users can call it for debugging purposes.